### ISSUE TITLE
CJ4 Lights update

### ISSUE THREAD

**Issue by ToddLorey** [Simstrumentation/Air-Manager#2]
I'm not that well versed in the ways of Github forking and pull requests so I'll just post this here.  I made a few edits to the lights panel that get the seatbelt, safety and pulse lights to work.  They do use some MobiFlight events.

````
--******************************************************************************************
-- ******************Cessna Citation CJ4 Light Control Panel****************************
--******************************************************************************************
--
-- Created and coded by Joe "Crunchmeister" Gilker
-- LIGHT DIMMER POTENTIOMETERS CURRENTLY INOP
-- Updated 8/29/21 by Todd "Toddimus831" Lorey
-- Added seatbelt, safety and pulse light functionality
--******************************************************************************************

--IMPORT ASSETS
img_add_fullscreen("bg.png")
fail_snd = sound_add("beepfail.wav")
click_snd=sound_add("Asobo_CJ4_WT_PC_75.wav")

--------SET UP INDIVIDUAL LIGHT SWITCHES

-- BEACON LIGHT
function beacon_click_callback(position)
    if position == 0 then
        switch_set_position(beacon_id, 1)
        fs2020_event("TOGGLE_BEACON_LIGHTS")
    elseif position == 1 then
        switch_set_position(beacon_id, 0)
        fs2020_event("TOGGLE_BEACON_LIGHTS")
    end
    sound_play(click_snd)
end

beacon_id = switch_add("beacon_off.png", "beacon_on.png", 110, 227, 65, 50, beacon_click_callback)

function new_beacon_pos(beacon_on)
    if beacon_on  then
        switch_set_position(beacon_id, 1)
    else
        switch_set_position(beacon_id, 0)
    end
end    
fs2020_variable_subscribe("LIGHT BEACON", "Bool", new_beacon_pos)
-- END BEACON LIGHT

-- NAV LIGHT
function nav_click_callback(position)
    if position == 0 then
        switch_set_position(nav_id, 1)
        fs2020_event("TOGGLE_NAV_LIGHTS")
    elseif position == 1 then
        switch_set_position(nav_id, 0)
        fs2020_event("TOGGLE_NAV_LIGHTS")
    end
    sound_play(click_snd)
end

nav_id = switch_add("nav_off.png", "nav_on.png", 207, 227, 65, 50, nav_click_callback)

function new_nav_pos(nav_on)
    if nav_on   then
        switch_set_position(nav_id, 1)
    else
        switch_set_position(nav_id, 0)
    end

end    
fs2020_variable_subscribe("LIGHT NAV", "Bool", new_nav_pos)
-- END NAV LIGHT

-- STROBE LIGHT
function strobe_click_callback(position)
    if position == 0 then
        switch_set_position(strobe_id, 1)
        fs2020_event("STROBES_TOGGLE")
    elseif position == 1 then
        switch_set_position(strobe_id, 0)
        fs2020_event("STROBES_TOGGLE")
    end
    sound_play(click_snd)
end

strobe_id = switch_add("strobe_off.png", "strobe_on.png", 306, 227, 65, 50, strobe_click_callback)

function new_strobe_pos(nav_on)
    if nav_on  then
        switch_set_position(strobe_id, 1)
    else
        switch_set_position(strobe_id, 0)
    end
end    
fs2020_variable_subscribe("LIGHT STROBE ON", "Bool", new_strobe_pos)
-- END STROBE LIGHT

-- TAXI LIGHT
function taxi_click_callback(position)
    if position == 0 then
        switch_set_position(taxi_id, 1)
        fs2020_event("TOGGLE_TAXI_LIGHTS")
    elseif position == 1 then
        switch_set_position(taxi_id, 0)
        fs2020_event("TOGGLE_TAXI_LIGHTS")
    end
    sound_play(click_snd)
end

taxi_id = switch_add("taxi_off.png", "taxi_on.png", 110, 306, 65, 50, taxi_click_callback)

function new_taxi_pos(taxi_on)
    if taxi_on  then
        switch_set_position(taxi_id, 1)
    else
        switch_set_position(taxi_id, 0)
    end
end    
fs2020_variable_subscribe("LIGHT TAXI ON", "Bool", new_taxi_pos)
-- END TAXI LIGHT

-- LANDING LIGHT
function land_click_callback(position)
    if position == 0 then
        switch_set_position(land_id, 1)
        fs2020_event("LANDING_LIGHTS_TOGGLE")
    elseif position == 1 then
        switch_set_position(land_id, 0)
        fs2020_event("LANDING_LIGHTS_TOGGLE")
    end
    sound_play(click_snd)
end

land_id = switch_add("land_off.png", "land_on.png", 208, 306, 65, 50, land_click_callback)

function new_land_pos(land_on)
    if land_on  then
        switch_set_position(land_id, 1)
    else
        switch_set_position(land_id, 0)
    end
end    
fs2020_variable_subscribe("LIGHT LANDING ON", "Bool", new_land_pos)
-- END LANDING LIGHT

-- LOGO LIGHT
function logo_click_callback(position)
    if position == 0 then
        switch_set_position(logo_id, 1)
        fs2020_event("TOGGLE_LOGO_LIGHTS")
    elseif position == 1 then
        switch_set_position(logo_id, 0)
        fs2020_event("TOGGLE_LOGO_LIGHTS")
    end
    sound_play(click_snd)
end

logo_id = switch_add("logo_off.png", "logo_on.png", 306, 306, 65, 50, logo_click_callback)

function new_logo_pos(logo_on)
    if logo_on then
        switch_set_position(logo_id, 1)
    else 
        switch_set_position(logo_id, 0)
    end
end    
fs2020_variable_subscribe("LIGHT LOGO ON", "Bool", new_logo_pos)
-- END LOGO LIGHT

-- BELT LIGHT
function belt_click_callback(position)
    if position == 0 then
        switch_set_position(belt_id, 1)
        --fs2020_event("CABIN_SEATBELTS_ALERT_SWITCH_TOGGLE")
        fs2020_event("Mobiflight.WT_CJ4_SEATBELT_LIGHT_TOGGLE")
    elseif position == 1 then
        switch_set_position(belt_id, 0)
        --fs2020_event("CABIN_SEATBELTS_ALERT_SWITCH_TOGGLE")
        fs2020_event("Mobiflight.WT_CJ4_SEATBELT_LIGHT_TOGGLE")
     end
    sound_play(fail_snd)
end

belt_id = switch_add("belt_off.png", "belt_on.png", 434, 227, 65, 50, belt_click_callback)

function new_belt_pos(sw_on)
    if sw_on == false then
        switch_set_position(belt_id, 0)
    elseif  sw_on == true then
        switch_set_position(belt_id, 1)
    end
end    
fs2020_variable_subscribe("SEATBELT_LIGHT_ON", "Bool", new_belt_pos)
-- END BELT LIGHT

-- SAFETY LIGHT
function safety_click_callback(position)
    if position == 0 then
        switch_set_position(safety_id, 1)
        fs2020_event("Mobiflight.WT_CJ4_SAFETY_LIGHT_TOGGLE")
    elseif position == 1 then
        switch_set_position(safety_id, 0)
        fs2020_event("Mobiflight.WT_CJ4_SAFETY_LIGHT_TOGGLE")
    end
    sound_play(fail_snd)
end

safety_id = switch_add("safety_off.png", "safety_on.png", 531, 227, 65, 50, safety_click_callback)

function new_safety_pos(sw_on)
    if sw_on == false then
        switch_set_position(safety_id, 0)
    elseif  sw_on == true then
        switch_set_position(safety_id, 1)
    end
end    
fs2020_variable_subscribe("SAFETY_LIGHT_ON", "Bool", new_safety_pos)
-- END SAFETY LIGHT

-- TCAS 
function tcas_click_callback(position)
    if position == 0 then
        switch_set_position(tcas_id, 1)
        fs2020_event("PULSE_LIGHTS_TOGGLE")
    elseif position == 1 then
        switch_set_position(tcas_id, 0)
        fs2020_event("PULSE_LIGHTS_TOGGLE")
    end
    sound_play(fail_snd)
end

tcas_id = switch_add("tcas_on.png", "tcas_off.png", 433, 306, 65, 50, tcas_click_callback)

function new_tcas_pos(sw_on)
    if sw_on == 0 then
        switch_set_position(tcas_id, 0)
    elseif  sw_on == 1 then
        switch_set_position(tcas_id, 1)
    end
end    
-- END TCAS

-- ON SWITCH

function on_click_callback(position)
    if position == 0 then
        switch_set_position(on_id, 1)
        --INOP
        fs2020_event("TOGGLE_RECOGNITION_LIGHTS")
    elseif position == 1 then
        switch_set_position(on_id, 0)
        --INOP
        fs2020_event("TOGGLE_RECOGNITION_LIGHTS")
    end
    sound_play(click_snd)
end

on_id = switch_add("on_off.png", "on_on.png", 531, 306, 65, 50, on_click_callback)

function new_on_pos(on_on)
    if on_on  then
        switch_set_position(on_id, 1)
    else 
        switch_set_position(on_id, 0)
    end
end    
fs2020_variable_subscribe("LIGHT RECOGNITION ON", "Bool", new_on_pos)
-- END ON  SWITCH
````

**Comment by Crunchmeister1** [comment]
What changes did you make to the original script? I'm struggling here to find what you changed.

**Comment by ToddLorey** [comment]
Just the part below.  I'm trying out a pull request right now actually but can't seem to figure out how to do it via the web interface and post it to the repository.
```

-- BELT LIGHT
function belt_click_callback(position)
    if position == 0 then
        switch_set_position(belt_id, 1)
        --fs2020_event("CABIN_SEATBELTS_ALERT_SWITCH_TOGGLE")
        fs2020_event("Mobiflight.WT_CJ4_SEATBELT_LIGHT_TOGGLE")
    elseif position == 1 then
        switch_set_position(belt_id, 0)
        --fs2020_event("CABIN_SEATBELTS_ALERT_SWITCH_TOGGLE")
        fs2020_event("Mobiflight.WT_CJ4_SEATBELT_LIGHT_TOGGLE")
     end
    sound_play(fail_snd)
end

belt_id = switch_add("belt_off.png", "belt_on.png", 434, 227, 65, 50, belt_click_callback)

function new_belt_pos(sw_on)
    if sw_on == false then
        switch_set_position(belt_id, 0)
    elseif  sw_on == true then
        switch_set_position(belt_id, 1)
    end
end    
fs2020_variable_subscribe("SEATBELT_LIGHT_ON", "Bool", new_belt_pos)
-- END BELT LIGHT

-- SAFETY LIGHT
function safety_click_callback(position)
    if position == 0 then
        switch_set_position(safety_id, 1)
        fs2020_event("Mobiflight.WT_CJ4_SAFETY_LIGHT_TOGGLE")
    elseif position == 1 then
        switch_set_position(safety_id, 0)
        fs2020_event("Mobiflight.WT_CJ4_SAFETY_LIGHT_TOGGLE")
    end
    sound_play(fail_snd)
end

safety_id = switch_add("safety_off.png", "safety_on.png", 531, 227, 65, 50, safety_click_callback)

function new_safety_pos(sw_on)
    if sw_on == false then
        switch_set_position(safety_id, 0)
    elseif  sw_on == true then
        switch_set_position(safety_id, 1)
    end
end    
fs2020_variable_subscribe("SAFETY_LIGHT_ON", "Bool", new_safety_pos)
-- END SAFETY LIGHT

-- TCAS 
function tcas_click_callback(position)
    if position == 0 then
        switch_set_position(tcas_id, 1)
        fs2020_event("PULSE_LIGHTS_TOGGLE")
    elseif position == 1 then
        switch_set_position(tcas_id, 0)
        fs2020_event("PULSE_LIGHTS_TOGGLE")
    end
    sound_play(fail_snd)
end

tcas_id = switch_add("tcas_on.png", "tcas_off.png", 433, 306, 65, 50, tcas_click_callback)

function new_tcas_pos(sw_on)
    if sw_on == 0 then
        switch_set_position(tcas_id, 0)
    elseif  sw_on == 1 then
        switch_set_position(tcas_id, 1)
    end
end    
-- END TCAS
```

**Comment by ToddLorey** [comment]
Have a look here, if you can access it:

https://github.com/ToddLorey/Air-Manager/pull/1/files

**Comment by rob2791** [comment]
Thanks Todd, I'm working on the panel now using your Joe's Original, your
updates and now I'm adding the variable subscribes so the button light
indicators work if you use the in-game buttons or external controllers.
You don't have to do a Git pull for this right now.
Thanks for the help and keep the updates coming. Joe and I just started
this Github and still have to organize some things.
Thanks,
Rob Verdon

**Comment by ToddLorey** [comment]
Cool!  Glad it helped.

**Comment by ToddLorey** [comment]
I'm pretty sure the way I have it makes the lights work whether you press them on the Air Manager panel or in-game.  I think I changed the variable subscribe, the event to send as well as the on/off state of the buttons in AM.  Been using it this way for a few weeks actually.

**Comment by Crunchmeister1** [comment]
Good stuff.

We'll be looking into removing Mobiflight events where / if possible, as that's the requirement to have instruments included in the AM community store. ATM, the CJ4 uses a ton of Mobiflight events.  We'll be chipping away at them when we have time to get them as "vanilla" SDK / SimConnect as possible.

**Comment by ToddLorey** [comment]
Yeah, agreed.  I'd like to be MobiFlight independent as well.  Not that it's really that bad, it's just that I'd prefer to have less stuff running in the background.  Let's hope they open up the SDK further.

**Comment by rob2791** [comment]
Thanks to Todd and Joe, this panel is close to 100%. The only thing left is Dials and I've just run out of time tonight. 
I was able to remove all the MobiFlight code and get all the buttons to subscribe to variables, so now if you click the button in game cockpit (or use a External Controller like a Yoke with option buttons such as a Logitech or Honeycomb) the AM panels lights operate correctly. 

I'm going to close this issue. I'll look into the dials next, but if someone else wants to tackle the Dials just let me know.
I uploaded tonights revision.
https://github.com/Simstrumentation/Air-Manager/tree/main/Instruments/Cessena_Citation_CJ4/CJ4-Light_Panel

**Comment by ToddLorey** [comment]
I was just messing with the sections I tweaked and I'm not sure it's all that clean.  The belt and safety lights don't update in the AM panel if you press them in-game.  And I had swapped the TCAS and Pulse lights stuff.

Sorry to hand off buggy mods!  Maybe you fixed them?  I'm a bit of a rookie with Lua and AM.

**Comment by ToddLorey** [comment]
I just saw that you updated it.  Looks like you found and fixed my bugs anyways!  ;)
Well done sir!

**Comment by rob2791** [comment]
Try the latest revision and let me know! 
We welcome the help! 
Joe and I have learned just by reverse engineering other AM panels and reading forums. I don’t think either of us consider ourselves experts. I just really want to fly with touchscreens in lieu of building cockpits and be stuck to only one plane.  :)

**Comment by ToddLorey** [comment]
I'll have a look at the dials now.  I'll try the big one in the middle first to keep it simple. 
A long time ago, I made some tweaks to the store's Mooney engine gauges to re-scale the EGT, I think it was.  I'll look back at that and maybe we can add it to the repo.

**Comment by rob2791** [comment]
Yeah go ahead! Even if you get the “dial” part to work in AirManager to turn and look good that’s a step in the right direction. Usually the graphics are 3/4 of the work. I’m certainly no artist and right now were just using the game graphics.

**Comment by ToddLorey** [comment]
I had a look and that seems like the tricky part.  I need to learn skinman.  I watched part of a tutorial on youtube.  I have much to learn before I tackle the knobs.  That ain't going to happen tonight.  :)

It makes me respect what you and Joe (and others) have done all the more!

I had a look through the Air Manager Community instrument that I adapted for the Mooney engine gauges.  I had stripped out the X-plane references and replaced/tweaked the code to work with MSFS.  The key change to make it match the sim values was scaling the EGT variable by 0.741.  I took empirical data at various throttle/mixture settings and did an Excel line fit to find the scaling factor.  I can clean it up and post the code if you guys want to add it to this repo.  I also adapted the Mooney Fuel Panel to MSFS as well.
